File: frontend/src/app/viewmodel/tour-log.viewmodel.ts
// MVVM: log form state + validation between View and Model.
import { TourLog } from '../models/tour-log';

export class TourLogViewModel {
  id?: number;
  logDate: string = new Date().toISOString().slice(0, 10);
  comment: string = '';
  difficulty: number = 1;
  totalDistance: number = 0;
  totalTimeMinutes: number = 0;
  rating: number = 5;
  logDetails: string = '';
  submitted: boolean = false;

  static from(log: TourLog): TourLogViewModel {
    const vm = new TourLogViewModel();
    vm.id = log.id;
    vm.logDate = log.logDate;
    vm.comment = log.comment;
    vm.difficulty = log.difficulty;
    vm.totalDistance = log.totalDistance;
    vm.totalTimeMinutes = log.totalTimeMinutes;
    vm.rating = log.rating;
    vm.logDetails = log.logDetails ?? '';
    return vm;
  }

  isValid(): boolean {
    return (
      this.logDate.length > 0 &&
      this.comment.trim().length > 0 &&
      this.difficulty >= 1 && this.difficulty <= 5 &&
      this.rating >= 0 && this.rating <= 10
    );
  }

  toLog(): TourLog {
    return {
      id: this.id,
      logDate: this.logDate,
      comment: this.comment,
      difficulty: this.difficulty,
      totalDistance: this.totalDistance,
      totalTimeMinutes: this.totalTimeMinutes,
      rating: this.rating,
      logDetails: this.logDetails.trim() ? this.logDetails : undefined,
    };
  }
}

File: backend/src/main/java/tour_planner_lamthi_mehmeti/model/TourLogViewModel.java
package tour_planner_lamthi_mehmeti.model;

import java.time.LocalDate;

/**
 * "View model" representation of a {@link TourLog}.
 *
 * <p>This class demonstrates the <b>MVVM pattern on the backend side</b>: it
 * exposes exactly the fields the UI needs, stripped of JPA annotations,
 * lifecycle callbacks, and the parent {@code Tour} relationship (which would
 * otherwise cause Jackson to serialize a huge graph). Keeping the view model
 * separate from the JPA entity lets us evolve the database schema without
 * breaking the JSON wire format, and vice-versa.
 *
 * <p>Currently used as a lightweight projection for scenarios where the full
 * entity would be overkill or would introduce circular references in JSON
 * serialization.
 */
public class TourLogViewModel {

    /** Primary key of the underlying log. */
    private Long id;
    /** Date the tour log was recorded. */
    private LocalDate logDate;
    /** Free-text remarks. */
    private String comment;
    /** Subjective 1–10 difficulty rating. */
    private Integer difficulty;
    /** Total distance actually walked / driven on this attempt (km). */
    private Double totalDistance;
    /** Elapsed time in whole minutes. */
    private Integer totalTimeMinutes;
    /** Subjective 1–10 rating of the experience. */
    private Integer rating;
    /** Long-form trip report. */
    private String logDetails;

    /** Default constructor required by Jackson for JSON deserialization. */
    public TourLogViewModel() {}

    /**
     * Copy-constructor that projects a {@link TourLog} entity into this
     * simpler view-model shape.
     *
     * @param tourLog source entity
     */
    public TourLogViewModel(TourLog tourLog) {
        this.id = tourLog.getId();
        this.logDate = tourLog.getLogDate();
        this.comment = tourLog.getComment();
        this.difficulty = tourLog.getDifficulty();
        this.totalDistance = tourLog.getTotalDistance();
        this.totalTimeMinutes = tourLog.getTotalTimeMinutes();
        this.rating = tourLog.getRating();
        this.logDetails = tourLog.getLogDetails();
    }

    /** Static factory — reads better at call sites than {@code new TourLogViewModel(log)}. */
    public static TourLogViewModel fromTourLog(TourLog tourLog) {
        return new TourLogViewModel(tourLog);
    }

    public Long getId() { return id; }
    public void setId(Long id) { this.id = id; }

    public LocalDate getLogDate() { return logDate; }
    public void setLogDate(LocalDate logDate) { this.logDate = logDate; }

    public String getComment() { return comment; }
    public void setComment(String comment) { this.comment = comment; }

    public Integer getDifficulty() { return difficulty; }
    public void setDifficulty(Integer difficulty) { this.difficulty = difficulty; }

    public Double getTotalDistance() { return totalDistance; }
    public void setTotalDistance(Double totalDistance) { this.totalDistance = totalDistance; }

    public Integer getTotalTimeMinutes() { return totalTimeMinutes; }
    public void setTotalTimeMinutes(Integer totalTimeMinutes) { this.totalTimeMinutes = totalTimeMinutes; }

    public Integer getRating() { return rating; }
    public void setRating(Integer rating) { this.rating = rating; }

    public String getLogDetails() { return logDetails; }
    public void setLogDetails(String logDetails) { this.logDetails = logDetails; }
}
